### PR TITLE
removing legacy profiler arg

### PR DESCRIPTION
**Pull request by Tshimanga**
## What does this PR do?

Clean up PR that removes the legacy profiler argument `profiled_functions`. This is due for removal for v1.5

If we didn't discuss your PR in Github issues there's a high chance it will not be merged.
We discussed this in the slack channel.

The following links the related issue to the PR (https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
-->

Fixes #\<issue_number>

## Before submitting

- [ x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ x] Did you write any **new necessary tests**? (not for typos and docs)
- [ x] Did you verify new and **existing tests pass** locally with your changes?
- [ x] Did you list all the **breaking changes** introduced by this pull request?
- [x ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/profiler/pytorch.py [hunk 1]
```diff
@@ -277,14 +276,12 @@ def __init__(
...
-
         self._group_by_input_shapes = group_by_input_shapes and profiler_kwargs.get("record_shapes", False)
         self._emit_nvtx = emit_nvtx
         self._export_to_chrome = export_to_chrome
         self._row_limit = row_limit
         self._sort_by_key = sort_by_key or f"{'cuda' if profiler_kwargs.get('use_cuda', False) else 'cpu'}_time_total"
-        self._user_record_functions = record_functions
+        self._user_record_functions = set() if record_functions is None else record_functions
         self._record_functions_start = self._user_record_functions | self.START_RECORD_FUNCTIONS
         self._record_functions = self._user_record_functions | self.RECORD_FUNCTIONS
```
Comment (Tshimanga) [inline pytorch_lightning/profiler/pytorch.py:286]: Tests were failing due to `|` operator not supporting `NoneType | Set`

File: tests/profiler/test_profiler.py [hunk 2]
```diff
@@ -258,12 +258,6 @@ def test_pytorch_profiler_describe(pytorch_profiler):
     assert len(data) > 0
 
 
-def test_pytorch_profiler_raises(pytorch_profiler):
-    """Ensure errors are raised where expected."""
-    with pytest.raises(MisconfigurationException, match="profiled_functions` and `PyTorchProfiler.record"):
-        PyTorchProfiler(profiled_functions=["a"], record_functions=["b"])
-
-
```
Comment (Tshimanga) [inline tests/profiler/test_profiler.py:266]: I wasn't entirely sure if removing this test entirely was the right call. After removing `profiled_functions` this raises a `TypeError` instead since `record_functions` expects a `Set` not a `List`. I could instead just update the test to check for that raised `TypeError` if desired.

File: CHANGELOG.md [hunk 3]
```diff
@@ -231,6 +231,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Removed `teardown` from `ParallelPlugin` ([#8943](https://github.com/PyTorchLightning/pytorch-lightning/pull/8943))
 
 
+- Removed `profiled_functions` argument from `PyTorchProfiler` ([#9178](https://github.com/PyTorchLightning/pytorch-lightning/pull/9178))
```
Comment (awaelchli) [inline CHANGELOG.md:234]:
```suggestion
- Removed deprecated `profiled_functions` argument from `PyTorchProfiler` ([#9178](https://github.com/PyTorchLightning/pytorch-lightning/pull/9178))
```

File: pytorch_lightning/core/hooks.py [hunk 4]
```diff
@@ -20,7 +20,7 @@
 
 from pytorch_lightning.utilities import move_data_to_device
 from pytorch_lightning.utilities.types import EVAL_DATALOADERS, STEP_OUTPUT, TRAIN_DATALOADERS
-from pytorch_lightning.utilities.warnings import rank_zero_deprecation, rank_zero_warn
```
Comment (awaelchli) [inline pytorch_lightning/core/hooks.py:23]: why this change? :)

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/core/hooks.py (hunk 4 above)
Comment (Tshimanga) [reply pytorch_lightning/core/hooks.py:23]: the linter in the `pre-commit.ci` caught an unused import there 😅

File: CHANGELOG.md (hunk 3 above)
Comment (Tshimanga) [reply CHANGELOG.md]: this commit-able suggestion feature is quite nice :)

**Review by tchaton — APPROVED** [review]
Great ! Thanks

File: tests/profiler/test_profiler.py (hunk 2 above)
Comment (tchaton) [reply tests/profiler/test_profiler.py:266]: All good !

File: CHANGELOG.md [hunk 5]
```diff
@@ -231,6 +231,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Removed `teardown` from `ParallelPlugin` ([#8943](https://github.com/PyTorchLightning/pytorch-lightning/pull/8943))
 
 
+- Removed deprecated `profiled_functions` argument from `PyTorchProfiler` ([#9178](https://github.com/PyTorchLightning/pytorch-lightning/pull/9178))
```
Comment (Borda) [inline CHANGELOG.md:234]:
```suggestion
- Removed deprecated `profiled_functions` and `record_functions` arguments from `PyTorchProfiler` ([#9178](https://github.com/PyTorchLightning/pytorch-lightning/pull/9178))
```

File: pytorch_lightning/profiler/pytorch.py [hunk 6]
```diff
@@ -277,14 +276,12 @@ def __init__(
...
 
-        record_functions = self.__deprecation_check(profiled_functions, record_functions)
-
         self._group_by_input_shapes = group_by_input_shapes and profiler_kwargs.get("record_shapes", False)
         self._emit_nvtx = emit_nvtx
         self._export_to_chrome = export_to_chrome
         self._row_limit = row_limit
         self._sort_by_key = sort_by_key or f"{'cuda' if profiler_kwargs.get('use_cuda', False) else 'cpu'}_time_total"
-        self._user_record_functions = record_functions
+        self._user_record_functions = set() if record_functions is None else record_functions
```
Comment (carmocca) [inline pytorch_lightning/profiler/pytorch.py:284]:
```suggestion
        self._user_record_functions = record_functions or set()
```

**Review by carmocca — APPROVED** [review]

File: CHANGELOG.md (hunk 5 above)
Comment (Tshimanga) [reply CHANGELOG.md:234]: @Borda The `record_functions` argument doesn't seem to be deprecated or due for removal. Did I miss something?

File: pytorch_lightning/profiler/pytorch.py (hunk 6 above)
Comment (Tshimanga) [reply pytorch_lightning/profiler/pytorch.py]: @carmocca ah right, I always forget about that `or` operator. Thanks, I'll include this change in my next contribution 👍🏽

File: CHANGELOG.md (hunk 5 above)
Comment (Borda) [reply CHANGELOG.md:234]: you were removing the warning which was referring to it...
